### PR TITLE
Fix SSA subtitles ignore the aspect ratio

### PR DESCRIPTION
**Pull request by Deathspike**
**Changes**
This PR changes the way `htmlVideoPlayer` applies aspect ratio. Instead of defaulting to applying `object-fit: ratio` to the video element, it now depends on the aspect ration in question. The behavior is unchanged for `aspect-ratio: fill` as there is currently no way with `libass` to scale according to said aspect ratio, but the behavior for `aspect-ratio: cover` now works as intended. The change resizes and moves the video element rather than relying on `object-fit`, so that `libass` can determine the proper size for its own rendering canvas. We then move said canvas along with the video element.

**Screenshots**
[Before] Movie screenshot with aspect-ratio auto:
before-auto

[Before] Movie screenshot with aspect-ratio cover:
before-fill

[After] Movie screenshot with aspect-ratio cover:
fixed

**Issues**
* https://github.com/jellyfin/jellyfin-web/issues/3599

### CODE REVIEW

**Review by dmitrylyzo — COMMENTED** [review]
Hard to say, but yes, this can probably only be done here (in the code of the library user).

Can't check this on webOS 1/2 - the emulators just stretch the video vertically :man_shrugging:

File: src/plugins/htmlVideoPlayer/plugin.js [hunk 1]
```diff
@@ -265,13 +265,53 @@ function tryRemoveElement(elem) {
...
+            switch (this._currentAspectRatio) {
+                case 'fill': {
+                    this.#mediaElement.setAttribute('style', 'object-fit: fill');
+                    if (subtitleCanvas) subtitleCanvas.style.removeProperty('margin-left');
+                    break;
+                }
+                case 'cover': {
+                    const coverWidth = Math.round(window.innerHeight / this.#mediaElement.videoHeight * this.#mediaElement.videoWidth);
+                    const centerLeft = Math.round((coverWidth - window.innerWidth) / 2);
+                    this.#mediaElement.setAttribute('style', `margin-left: -${centerLeft}px !important; width: ${coverWidth}px`);
```
[the remark this answers is not in the document]
Comment (Deathspike) [reply src/plugins/htmlVideoPlayer/plugin.js:300]: Wait, what? Chrome 79? How many past versions is `jellyfin-web` supposed to support?

File: src/plugins/htmlVideoPlayer/plugin.js [hunk 2]
```diff
@@ -265,13 +265,53 @@ function tryRemoveElement(elem) {
...
+                }
+                default: {
+                    this.#mediaElement.removeAttribute('style');
+                    if (subtitleCanvas) subtitleCanvas.style.removeProperty('margin-left');
+                    break;
+                }
+            }
+            if (this.#currentSubtitlesOctopus) {
+                this.#currentSubtitlesOctopus.resetRenderAheadCache(false);
+            }
```
[the remark this answers is not in the document]
Comment (Deathspike) [reply src/plugins/htmlVideoPlayer/plugin.js:312]: The frames rendered ahead of time are rendered for the old element size. When you switch from a small element size to a large element size, the frame rendered at the low resolution is stretched to the large element size, making everything appear very blurry. This is why the cache needs to be reset.

File: src/plugins/htmlVideoPlayer/plugin.js [hunk 3]
```diff
@@ -265,13 +265,53 @@ function tryRemoveElement(elem) {
...
+                case 'fill': {
+                    this.#mediaElement.setAttribute('style', 'object-fit: fill');
+                    if (subtitleCanvas) subtitleCanvas.style.removeProperty('margin-left');
+                    break;
+                }
+                case 'cover': {
+                    const coverWidth = Math.round(window.innerHeight / this.#mediaElement.videoHeight * this.#mediaElement.videoWidth);
+                    const centerLeft = Math.round((coverWidth - window.innerWidth) / 2);
+                    this.#mediaElement.setAttribute('style', `margin-left: -${centerLeft}px !important; width: ${coverWidth}px`);
+                    if (subtitleCanvas) subtitleCanvas.style.marginLeft = `-${centerLeft}px`;
```
[the remark this answers is not in the document]
Comment (Deathspike) [reply src/plugins/htmlVideoPlayer/plugin.js:301]: Can you explain this point a little further? I haven't seen any issues in my testing with a narrow window. How did you test this and what occurred?

File: src/plugins/htmlVideoPlayer/plugin.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (dmitrylyzo) [reply src/plugins/htmlVideoPlayer/plugin.js:300]: Chrome 27 :p (we use it as a rough equivalent of WebKit used by Tizen2.x and webOS 1/2).

TVs have old web engines. For example, Tizen 3 is Chrome 47, webOS 3 is Chrome 38
https://developer.samsung.com/smarttv/develop/specifications/web-engine-specifications.html
https://webostv.developer.lge.com/discover/specifications/web-engine/

File: src/plugins/htmlVideoPlayer/plugin.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (dmitrylyzo) [reply src/plugins/htmlVideoPlayer/plugin.js:301]: Make the window wide, but lower height.
_1920x801 video in 1600x500 window_
wide_window
It looks the same as `Auto`.

On the other hand, I'm not sure if we need to fit the video by width in this case.

I have not tested the video in portrait orientation :thinking:

File: src/plugins/htmlVideoPlayer/plugin.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (dmitrylyzo) [reply src/plugins/htmlVideoPlayer/plugin.js:312]: It keeps 10s of render buffer - maybe too much.
I was going to add options for this and reduce the storage time to 3s.
OK, left this for now. At least the aspect ratio change doesn't happen too often.

File: src/plugins/htmlVideoPlayer/plugin.js [hunk 4]
```diff
@@ -265,13 +265,53 @@ function tryRemoveElement(elem) {
...
+                    break;
+                }
+                default: {
+                    this.#mediaElement.removeAttribute('style');
+                    if (subtitleCanvas) subtitleCanvas.style.removeProperty('margin-left');
+                    break;
+                }
+            }
+            if (this.#currentSubtitlesOctopus) {
+                this.#currentSubtitlesOctopus.resetRenderAheadCache(false);
```
Comment (dmitrylyzo) [inline src/plugins/htmlVideoPlayer/plugin.js:311]: **4K TV issue**
On Tizen 4 [there is no `ResizeObserver`](https://github.com/libass/JavascriptSubtitlesOctopus/blob/1e9d5cd2147420fe1485bd1b71fbc9ee51537641/src/subtitles-octopus.js#L200-L204), so JSO resize handler is not called.
Fixed by adding `if (typeof ResizeObserver === 'undefined') this.#currentSubtitlesOctopus.resize();` here.

**Pull request closed without merging**